src/rgw/driver/rados/rgw_rados.cc

```diff
@@ -3668,13 +3679,55 @@ class RGWRadosPutObj : public RGWHTTPStreamRWRequest::ReceiveCB
     return filter->process(std::move(bl), lofs);
   }
 
+  void set_in_stream_req(RGWRESTStreamRWRequest *req) override {
+    in_stream_req = req;
+  }
+
+  void make_sse_s3_key(bool *pause) override {
+    if (pause) *pause = true; // xxx: could this race with the spawn? (i.e. pause is set to curl after the spawn)
+
+    boost::asio::spawn(io_context, [&](boost::asio::yield_context yield) {
+      std::unique_ptr<rgw::sal::Bucket> bucket;
+      int ret = driver->load_bucket(dpp, obj.bucket, &bucket, yield);
+      if (ret < 0) {
+        ldpp_dout(dpp, 0) << "ERROR: failed to load bucket info for bucket="
+                          << obj.bucket << " ret=" << ret << dendl;
+
+        in_stream_req->cancel();
+        return;
+      }
+
+      std::unique_ptr<BlockCrypt> block_crypt;
+      std::string err_msg;
+      ret = handle_sse_s3_encryption(dpp, cct, yield, bucket.get(), obj.key, &err_msg, enc_attrs, &block_crypt, nullptr);
+      if (ret < 0) {
+        ldpp_dout(dpp, 0) << "ERROR: failed to handle sse-s3 encryption for bucket="
+                          << obj.bucket << " key=" << obj.key << " ret=" << ret << " err_msg=" << err_msg << dendl;
+
+        in_stream_req->cancel();
+        return;
+      }
+
+      if (block_crypt != nullptr) {
+        encrypt.reset(new RGWPutObj_BlockEncrypt(dpp, cct, filter, std::move(block_crypt), yield));
+        filter = &*encrypt;
+      }
+
+      in_stream_req->unpause_receive();
+    }, [] (std::exception_ptr eptr) {
+      if (eptr) std::rethrow_exception(eptr);
```

Comment:
I guess here we should only log?

Reply:
the functions called by this coroutine return error codes instead of throwing. if we only expect to get exceptions if something catastrophic happens (like bad_alloc), then i think it's best to rethrow them